presto-hive/src/main/java/com/facebook/presto/hive/HivePageSink.java

```diff
@@ -177,27 +186,60 @@ private void doAbort()
     @Override
     public CompletableFuture<?> appendPage(Page page)
     {
-        if (page.getPositionCount() == 0) {
-            return NOT_BLOCKED;
+        if (page.getPositionCount() > 0) {
+            // Must be wrapped in doAs entirely
+            // Implicit FileSystem initializations are possible in HiveRecordWriter#addRow or #createWriter
+            hdfsEnvironment.doAs(session.getUser(), () -> doAppend(page));
         }
 
-        // Must be wrapped in doAs entirely
-        // Implicit FileSystem initializations are possible in HiveRecordWriter#addRow or #createWriter
-        return hdfsEnvironment.doAs(session.getUser(), () -> doAppend(page));
+        return NOT_BLOCKED;
     }
 
-    private CompletableFuture<?> doAppend(Page page)
+    private void doAppend(Page page)
+    {
+        while (page.getPositionCount() > MAX_PAGE_POSITIONS) {
+            Page chunk = page.getRegion(0, MAX_PAGE_POSITIONS);
+            page = page.getRegion(MAX_PAGE_POSITIONS, page.getPositionCount() - MAX_PAGE_POSITIONS);
+            writePage(chunk);
+        }
+
+        writePage(page);
+    }
+
+    private void writePage(Page page)
     {
         int[] writerIndexes = getWriterIndexes(page);
 
-        Block[] dataBlocks = getDataBlocks(page);
+        // record which positions are used by which writer
         for (int position = 0; position < page.getPositionCount(); position++) {
             int writerIndex = writerIndexes[position];
-            HiveWriter writer = writers.get(writerIndex);
-            writer.addRow(dataBlocks, position);
+            tempWriterPositions.get(writerIndex).add(position);
         }
 
-        return NOT_BLOCKED;
+        // invoke the writers
+        Page dataPage = getDataPage(page);
+        IntSet writersUsed = new IntArraySet(writerIndexes);
+        for (IntIterator iterator = writersUsed.iterator(); iterator.hasNext(); ) {
+            int writerIndex = iterator.nextInt();
+            WriterPositions writerPositions = tempWriterPositions.get(writerIndex);
+            if (writerPositions.isEmpty()) {
+                continue;
+            }
+
+            // If write is partitioned across multiple writers, filter page using dictionary blocks
+            Page pageForWriter = dataPage;
+            if (writerPositions.size() != dataPage.getPositionCount()) {
+                Block[] blocks = new Block[dataPage.getChannelCount()];
+                for (int channel = 0; channel < dataPage.getChannelCount(); channel++) {
+                    blocks[channel] = new DictionaryBlock(writerPositions.size(), dataPage.getBlock(channel), writerPositions.getPositionsArray());
+                }
+                pageForWriter = new Page(writerPositions.size(), blocks);
+            }
+
+            HiveWriter writer = writers.get(writerIndex);
+            writer.append(pageForWriter);
+            writerPositions.clear();
```

Comment:
Add a comment to `writerPositions` that says it is a field instead of a local variable because we want to reduce allocation cost.